src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -399,124 +397,13 @@ fsm_resume_as_primary(Keeper *keeper)
- * in fsm_promote_standby too, where we could have to accept the fact that
- * there's no other node at the moment: we're doing a secondary ➜ single
- * transition after all.
- */
-static bool
-prepare_replication(Keeper *keeper, NodeState otherNodeState)
-{
-	KeeperConfig *config = &(keeper->config);
-	Monitor *monitor = &(keeper->monitor);
-	LocalPostgresServer *postgres = &(keeper->postgres);
-	PostgresSetup *pgSetup = &(postgres->postgresSetup);
-
-	int nodeIndex = 0;
-
-	/*
-	 * When using the monitor, now is the time to fetch the otherNode hostname
-	 * and port from it. When the monitor is disabled, it is expected that the
-	 * information has been already filled in keeper->otherNode; see
-	 * `keeper_cli_fsm_assign' for an example of that.
-	 */
-	if (!config->monitorDisabled)
-	{
-		char *host = config->nodename;
-		int port = pgSetup->pgport;
-
-		if (!monitor_get_other_nodes(monitor, host, port,
-									 otherNodeState,
-									 &(keeper->otherNodes)))
-		{
-			/* errors have already been logged */
-			return false;
-		}
-
-		if (keeper->otherNodes.count == 0)
-		{
-			if (otherNodeState == ANY_STATE)
-			{
-				log_warn("There's no other node for %s:%d", host, port);
-			}
-			else
-			{
-				/*
-				 * Should we warn about it really? it might be a replication
-				 * setting change that will impact synchronous_standby_names
-				 * and that's all.
-				 */
-				log_warn("There's no other node in state \"%s\" "
-						 "for node %s:%d",
-						 NodeStateToString(otherNodeState),
-						 host, port);
-			}
-		}
-	}
-
-	/*
-	 * Should we fail somewhere in this loop, we return false and fail the
-	 * whole transition. The transition is going to be tried again, and we are
-	 * going to try and add HBA entries and create replication slots again.
-	 * Both operations succeed when their target entry already exists.
-	 *
-	 */
-	for (nodeIndex = 0; nodeIndex < keeper->otherNodes.count; nodeIndex++)
-	{
-		NodeAddress *otherNode = &(keeper->otherNodes.nodes[nodeIndex]);
-		char replicationSlotName[BUFSIZE] = { 0 };
-
-		log_info("Preparing replication for standby node %d (%s:%d)",
-				 otherNode->nodeId, otherNode->host, otherNode->port);
-
-		if (!primary_add_standby_to_hba(postgres,
-										otherNode->host,
-										config->replication_password))
-		{
-			log_error("Failed to grant access to the standby %d (%s:%d) "
-					  "by adding relevant lines to pg_hba.conf for the standby "
-					  "hostname and user, see above for details",
-					  otherNode->nodeId, otherNode->host, otherNode->port);
-			return false;
-		}
-
-		if (!postgres_sprintf_replicationSlotName(otherNode->nodeId,
-												  replicationSlotName, BUFSIZE))
-		{
-			/* that's highly unlikely... */
-			log_error("Failed to snprintf replication slot name for node %d",
-					  otherNode->nodeId);
-			return false;
-		}
-
-		if (!primary_create_replication_slot(postgres, replicationSlotName))
```

Comment:
I'm a little confused. We're also removing `primary_create_replication_slot`, and I couldn't see where we create replication slots other than this (other than cli).

Shouldn't we need to create `primary_create_replication_slot ` at some point?

Reply:
That's a good point. The function that ensures that is badly named: `pgsql_replication_slot_drop_removed` is called by `postgres_replication_slot_drop_removed` and does both the `pg_drop_replication_slot` calls and also the `pg_create_physical_replication_slot` calls... Maybe this is a good time to rename that function.